### DISCUSSION TITLE
New Cookie handling since KC 24.0.0 blocks login without TLS usage. Any flag to disable "Secure" property in specific realms?

### DISCUSSION THREAD

**Discussion by psytester** [keycloak/keycloak#27592] [Q&A] [answered]
Is there any Realm setting or configuration possible to revert to old cookie handling?

In my **test** setup I'm using realms behind a reverse proxy which handles the TLS stuff
and I'm using realms for internal tools directly on keycloak internal port 8080 without TLS.

Since Keycloak 24.0.0 changed the cookie handling, I'm blocked now for the internal realms on plain HTTP port.
Because all Set-Cookies gets "Secure" now and after user login I fail into **error="cookie_not_found"**
```
DEBUG [org.keycloak.protocol.RestartLoginCookie:132] [executor-thread-6] KC_RESTART cookie doesn't exist
TRACE [org.keycloak.events:130] [executor-thread-6] type="LOGIN_ERROR", realmId="512987e8-4.....", clientId="null", userId="null", ipAddress="10.1.....", error="cookie_not_found", requestUri="http://keycloak:8080/auth/realms/realmname/login-actions/authenticate?session_code=....", cookies=[]
```

Up to Keycloak 23.0.7 only AUTH_SESSION_ID had "Secure" set:
```
HTTP GET http://keycloak:8080/auth/realms/realmname/protocol/openid-connect/auth?response_type=code&.....

Set-Cookie:
AUTH_SESSION_ID=6f48c6df-5cad-407e-.....; Version=1; Path=/auth/realms/realmname/; SameSite=None; Secure; HttpOnly, 
AUTH_SESSION_ID_LEGACY=6f48c6df-5cad-407e-.....; Version=1; Path=/auth/realms/realmname/; HttpOnly,
KC_RESTART=ey.....; Version=1; Path=/auth/realms/realmname/; HttpOnly',

HTTP POST Login url: http://keycloak:8080/auth/realms/realmname/login-actions/authenticate?session_code=.....
No Set-Cookie in response
```

Since Keycloak 24.0.0 / 24.0.1 all cookie get "Secure" set:
```
HTTP GET http://keycloak:8080/auth/realms/realmname/protocol/openid-connect/auth?response_type=code&......
'Set-Cookie': '
AUTH_SESSION_ID=8f9c26af-1a7a-4bbd-.....;Version=1;Path=/auth/realms/realmname/;Secure;HttpOnly;SameSite=None,
AUTH_SESSION_ID_LEGACY=8f9c26af-1a7a-4bbd-.....;Version=1;Path=/auth/realms/realmname/;Secure;HttpOnly,
KC_RESTART=ey.....;Version=1;Path=/auth/realms/realmname/;Secure;HttpOnly',


POST Login url: http://keycloak:8080/auth/realms/realmname/login-actions/authenticate?session_code=......
'Set-Cookie': 'KC_STATE_CHECKER=ey.....;Version=1;Path=/auth/realms/realmname/;Max-Age=300;Secure;HttpOnly',
```

From security point of view in a production env it's great to force it, but how to bypass this in a dev/test stage?

According to code, the sameSite is doing the job set or not "Secure"
https://github.com/stianst/keycloak/blob/main/services/src/main/java/org/keycloak/cookie/CookieSecureResolver.java#L19

**Comment by stianst** [comment]
I presume `keycloak` is not pointing to a local address? If so the default ssl-required for a realm is to require ssl for any non-local connections, and this is what will be setting the secure flag on the cookies.

**Reply by psytester** [reply, marked as answer]
project & keycloak is running in a relative close to production environment, but the internal tools are not TLS secured for my tests.
The applications are running on dedicated systems in a 10.x.x.x IP range and not on same localhost as my user browser.

Your hint about ssl-required was the key 👍 

Realm-->Realm Settings-->General-->Require SSL:
It was set `Require SSL = External requests` and worked with KC <= 23.0.7
it means `Users can interact with Keycloak without SSL so long as they stick to private IP addresses such as localhost, 127.0.0.1, 10.x.x.x, 192.168.x.x, and 172.16.x.x. If you try to access Keycloak without SSL from a non-private IP address, you will get an error.`

Although the network is using 10.x.x.x IPs it did not work since KC 24.0.x

I change to `Require SSL = None` because in my dev/test `Keycloak does not require SSL. This choice applies only in development when you are experimenting and do not plan to support this deployment.`

The Set-Cookie "Secure" is no longer set in that realm.

Issue is solved
